### PR TITLE
Add Base class for metrics

### LINKED ISSUE #1294: Metrics: Base Metric
## 🚀 Feature
Add a base class for proper metric implementation

### PR DESCRIPTION
**Pull request by justusschock**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Implements a base class for new metrics.
Fixes #1294 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

A review of the general interface would be nice, before I start to implement the necessary tests

### CODE REVIEW

File: pytorch_lightning/metrics/metric.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        """
+        raise NotImplementedError
+
+    def __call__(self, *args, **kwargs) -> torch.Tensor:
+        return sync_collections(super().__call__(*args, **kwargs),
+                                group=self.reduce_group,
+                                reduce_op=self.reduce_op)
+
+
+def sync_ddp(result: Union[torch.Tensor, numbers.Number],
```
Comment (Borda) [inline pytorch_lightning/metrics/metric.py]: this is public or private (do we want to expose it to users)?

Comment (justusschock) [reply pytorch_lightning/metrics/metric.py]: A mixture of both, usually I'd say, this should be private (hence it's not included in `__all__`).
More experienced users might however want to call it directly, which is why I didn't ad a `_` prefix.
But we can also make this private

Comment (Borda) [reply pytorch_lightning/metrics/metric.py]: I may add `_` so it clear that is not meant for general usage but you still can if you want...

File: pytorch_lightning/metrics/metric.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    elif isinstance(result, Mapping):
+        return elem_type({key: func(result[key]) for key in result})
+    elif isinstance(result, tuple) and hasattr(result, '_fields'):  # namedtuple
+        return elem_type(*(func(r) for r in result))
+    elif isinstance(result, Sequence) and not isinstance(result, str):
+        return elem_type([func(r) for r in result])
+    # not possible to reduce this type
+    else:
+        return result
```
Comment (Borda) [inline pytorch_lightning/metrics/metric.py]: this is not safe since all `result` assignment is under some conditions...
rather instead each return do result assignment and always return result

File: tests/metrics/base.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import pytest
+import torch
+import torch.distributed as dist
+
+import tests.base.utils as tutils
+from pytorch_lightning.metrics.metric import _sync_ddp, _sync_collections, BaseMetric
+
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2,
+                    'Not enough GPUs to test sync reduce')
```
Comment (Borda) [inline tests/metrics/base.py]:
```suggestion
@pytest.mark.skipif(torch.cuda.device_count() < 2, "test requires multi-GPU machine")
```

File: tests/metrics/base.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    number = 1.
+    reduced_number = _sync_ddp(number)
+    assert isinstance(reduced_number, torch.Tensor), 'When reducing a number we should get a tensor out'
+    assert reduced_number.item() == dist.get_world_size(), \
+        'Sync-Reduce does not work properly with DDP and Numbers'
+
+    dist.destroy_process_group()
+
+
+def test_sync_reduce_no_ddp():
```
Comment (Borda) [inline tests/metrics/base.py]:
```suggestion
def test_sync_reduce_simple():
```

File: tests/metrics/base.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+    expected_result = {
+        'a': torch.tensor([1. * factor]),
+        'b': [torch.tensor([2. * factor])],
+        'c': (torch.tensor([100. * factor]),),
+        'd': ntc(bar=torch.tensor([5. * factor])),
+        'e': torch.tensor([10. * factor]),
+        'f': 'this_is_a_dummy_str',
+        'g': torch.tensor([12. * factor])
+
```
Comment (Borda) [inline tests/metrics/base.py]:
```suggestion
        'g': torch.tensor([12. * factor]),
```

File: tests/metrics/base.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    tutils.set_random_master_port()
+
+    dist.init_process_group('gloo')
+
+    _sync_collections_test(True)
+
+    dist.destroy_process_group()
+
+
+def test_sync_collections_no_ddp():
```
Comment (Borda) [inline tests/metrics/base.py]:
```suggestion
def test_sync_collections_simple():
```

File: tests/metrics/base.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    assert metric_val.item() == expected, 'Invalid Value for reduction'
+
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2,
+                    'Not enough GPUs to test with ddp')
+def test_base_metric_ddp():
+    _test_base_metric(True)
+
+
+def test_base_metric_no_ddp():
```
Comment (Borda) [inline tests/metrics/base.py]:
```suggestion
def test_base_metric_simple():
```

File: pytorch_lightning/metrics/metric.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+import torch.distributed
+from torch.utils.data._utils.collate import np_str_obj_array_pattern
+
+__all__ = ['BaseMetric']
+
+
+class BaseMetric(torch.nn.Module, ABC):
+    def __init__(self, name: str,
+                 reduce_group: Optional[Any] = torch.distributed.group.WORLD,
+                 reduce_op: Optional[Any] = torch.distributed.ReduceOp.SUM):
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py]: mean is not available as a reduction op :) 
https://pytorch.org/docs/stable/distributed.html#torch.distributed.ReduceOp

File: pytorch_lightning/metrics/metric.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py]: this should be safe, since we just return the input variable as is, if there is nothing to sync...

File: pytorch_lightning/metrics/metric.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    @abstractmethod
+    def forward(self, *args, **kwargs) -> torch.Tensor:
+        """
+        Implements the actual metric computation.
+
+        Returns:
+            metric value
+
+        """
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py]: I know, but it doesn't matter if you put `pass` or `raise NotImplementedErro`r there and I found this to be more intuitive, but we can still change this, if you want...

[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/metrics/metric.py]: this line won't be ever tested so lover test coverage count... lol

**Comment by justusschock** [comment]
closed in favour of #1326

**Comment by Borda** [comment]
@justusschock what os the difference?

**Comment by justusschock** [comment]
@Borda the new version is cleaner and has more capabilities. This includes the changes, I had to make when implementing the scikit part. It was easier to make it as a separate or than to incorporate it into this PR
